ipaserver/plugins/timerule.py

```diff
@@ -0,0 +1,293 @@
+                        'subcomponent "%s".') % (sub.name, )
+                )
+        else:
+            root_logger.info(
+                'Found "{comp}" but only VEVENT component is '
+                'supported.'
+                .format(comp=sub.name))
+
+    # we WILL require DTSTART for VEVENTs
+    if 'DTSTART' not in comp.keys():
+        raise errors.ValidationError(
+            name=name,
+            error=_('DTSTART property is required in VEVENT.')
+            )
+
+    if 'DTEND' in comp.keys():
+        if 'DURATION' in comp.keys():
+            raise errors.ValidationError(
+                name=name,
+                error=_('Both DURATION and DTEND set in a VEVENT.')
+            )
+
+        if type(comp['DTSTART'].dt) != type(comp['DTEND'].dt):
+            raise errors.ValidationError(
+                name=name,
+                error=_('Different types of DTSTART and DTEND '
+                        'component in VEVENT.')
+                )
+
+    elif 'DURATION' in comp.keys() and isinstance(comp['DTSTART'].dt, date):
+        """
+        python-icalendar represents DURATION as datetime.timedelta. This,
+        in some cases, blocks us from checking whether it was originally
+        set correctly.
+
+        Example: If DTSTART has value of type DATE, DURATION should be set
+        only as dur-day or dur-week. However, DURATION:PT24H will evaluate
+        as timedelta(1)
+        """
+        if comp['DURATION'].dt.seconds:
+            raise errors.ValidationError(
+                name=name,
+                error=_('DURATION is not of type dur-day or dur-week '
+                        'when DTSTART value type is DATE.')
+                )
+
+
+@register()
+class timerule(LDAPObject):
+    """
+    Time Rule object
+    """
+    container_dn = api.env.container_timerules
+    object_name = _('Time Rule')
+    object_name_plural = _('Time Rules')
+    object_class = ['ipatimerule']
+    permission_filter_objectclasses = ['ipatimerule']
+    default_attributes = ['cn', 'description', 'accesstime', 'memberof']
+    attribute_members = {
+        'memberof': ['hbacrule']
+    }
+    label = _('Time Rules')
+    label_singular = _('Time Rule')
+    managed_permissions = {
+        'System: Read Time Rule': {
+            'replaces_global_anonymous_aci': True,
+            'ipapermbindruletype': 'all',
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermdefaultattr': {
+                'cn', 'objectclass', 'description', 'accesstime'
+            },
+        },
+        'System: Read Timerule Membership': {
+            'replaces_global_anonymous_aci': True,
+            'ipapermbindruletype': 'all',
+            'ipapermright': {'read', 'search', 'compare'},
+            'ipapermdefaultattr': {
+                'memberof',
+            },
+        },
+        'System: Add Time Rule': {
+            'ipapermright': {'add'},
+            'default_privileges': {'HBAC Administrator'},
```

Comment:
I'm curious for future, what if we will use timerules for sudo? then HBAC Administrator will be able to change timerules for Sudorules. IMO we should play safe and create a new Privilege "Time Rules Administrator". It will bother users to add both HBAC Administrator and Time Rule Administrator into the same priviledge, but it will be under user control. Otherwise we may get CVE :)